lib/uplink/bucket.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/storage/streams"
+	"storj.io/storj/pkg/storj"
+)
+
+// Bucket represents operations you can perform on a bucket
+type Bucket struct {
+	storj.Bucket
+
+	metainfo   *kvmetainfo.DB
+	streams    streams.Store
+	pathCipher storj.Cipher
+}
+
+// OpenObject returns an Object handle, if authorized.
+func (b *Bucket) OpenObject(ctx context.Context, path storj.Path) (o *Object, err error) {
+	defer mon.Task()(&ctx)(&err)
+	panic("TODO")
+}
+
+// UploadOptions controls options about uploading a new Object, if authorized.
+type UploadOptions struct {
+	// ContentType, if set, gives a MIME content-type for the Object.
+	ContentType string
+	// Metadata contains additional information about an Object. It can
+	// hold arbitrary textual fields and can be retrieved together with the
+	// Object. Field names can be at most 1024 bytes long. Field values are
+	// not individually limited in size, but the total of all metadata
+	// (fields and values) can not exceed 4 kiB.
+	Metadata map[string]string
+	// Expires is the time at which the new Object can expire (be deleted
+	// automatically from storage nodes).
+	Expires time.Time
+
+	// Volatile groups config values that are likely to change semantics
+	// or go away entirely between releases. Be careful when using them!
+	Volatile struct {
+		// DataCipher determines the ciphersuite to use for the Object's
+		// data encryption. If not set, the Bucket's default will be
+		// used.
+		DataCipher Cipher
+		// EncryptionBlockSize determines the unit size at which
+		// encryption is performed. See BucketConfig.EncryptionBlockSize
+		// for more information.
+		EncryptionBlockSize memory.Size
```

Comment:
We should use the `storj.EncryptionScheme` struct instead of separate `DataCipher` and `EncryptionBlockSize` fields.

Also in `ObjectMeta` and `BucketConfig`.

Reply:
See the comment on 5fbea44:

> It seems to make sense for each of these to have a RedundancyScheme, a DataCipher, and an EncryptionBlockSize. We could use storj.EncryptionScheme for the latter two, but we want a Cipher type where we can distinguish an unset value (0) from storj.Unencrypted (0). We could just use a *storj.EncryptionScheme and have the nil pointer mean to use defaults, but we would still have the same problem with PathCipher. It doesn't make sense for PathCipher to use storj.EncryptionScheme because it doesn't have a BlockSize. We could use uplink.Cipher for PathCipher and storj.EncryptionScheme for DataCipher, but the mismatch is mentally jarring.

Reply:
Why don't we just change the existing `storj.Cipher` to have `UnsetCipher` zero value instead of defining a brand new `uplink.Cipher`? It's a backward incompatible change, but we are anyway resetting the data in the complete V3 network with the release we are trying to push right now. So it's the best time to make this change.

Reply:
I'd be entirely ok with that, but we'd have to coordinate it with everybody and everything that might touch that field, and it might take longer. What if we made new `storj.CipherSuite` and `storj.EncryptionParameters` types and use those, and transition other things to using those over time?